File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length())
            return false;
        
        unordered_map<char,int> m1;
        
        for (auto const &c:s1) ++m1[c];
        for (int i = 0; i <= s2.length() - s1.length(); ++i) {
            unordered_map<char,int> m2;
            auto sstr = s2.substr(i,s1.length());
            for (auto const &c:sstr) ++m2[c];
            
            if (m1 == m2) return true;
        }
        return false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.rs
impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let (n1,n2) = (s1.len(), s2.len());
        if n1 > n2 { return false; }
        
        let (charr1,charr2):(Vec<char>,Vec<char>) = (s1.chars().collect(),s2.chars().collect());
        let (mut m1, mut m2) = (vec![0;26], vec![0;26]);
        
        for i in 0..n1 {
            m1[charr1[i] as usize - 'a' as usize] += 1;
            m2[charr2[i] as usize - 'a' as usize] += 1;
        }
        
        if m1 == m2 { return true; }
        
        for i in n1..n2 {
            m2[charr2[i] as usize - 'a' as usize] += 1;
            m2[charr2[i - n1] as usize - 'a' as usize] -= 1;
            if m1 == m2 { return true; }
        }
        
        false
    }
}